### DISCUSSION TITLE
ES7P003FGTF support

### DISCUSSION THREAD

**Discussion by rtek1000** [TG9541/stm8ef#465] [General]
Hello,

Do you know how to program the ES7P003?  Tools and IDE? I received a W1209 board with this microcontroller.

Maybe you can add the ES7P003FGTF on this page:

https://github.com/TG9541/stm8ef/wiki/W1209-Identifying-'compliant'-boards

ES7P003 (The board does not have Vcap)
![ES7P003](https://user-images.githubusercontent.com/8660811/229627296-3e41c706-0bfb-424e-99bb-cb57bcafb71c.png)

STM8S003
![STM8S003F3P6](https://user-images.githubusercontent.com/8660811/229627325-bb840d1f-952a-4247-9a96-5f1192ddbf1e.jpeg)

Datasheet (Automatically translated):
https://github.com/rtek1000/W1209-firmware-modified/blob/master/W1209/Doc/ES7P003FGTF_EN.pdf

Datasheet original:
https://github.com/rtek1000/W1209-firmware-modified/blob/master/W1209/Doc/ES7P003FGTF_CN_622163.pdf

**Comment by rtek1000** [comment]
Oh, I forgot to comment, the W3230 board also has an STM8S003F3P6.

Thanks.

**Comment by rtek1000** [comment]
I found the manufacturer's website, it seems that everything is available for download, from the compiler to the programmer (ES-Link 😂) and via the serial port.

https://www.essemi.com/index/product/detail?id=39

**Comment by rtek1000** [comment]
Sadly, QFN24 is also used on the W1209:

![image](https://i.stack.imgur.com/fKjRw.jpg)

[(EFM8SB20F16G-A)](https://br.mouser.com/ProductDetail/Silicon-Labs/EFM8SB20F16G-B-QFN24R?qs=OycAS1CGnlgZuuivzFrL3w%3D%3D)

https://electronics.stackexchange.com/questions/591867/can-anyone-identify-the-microcontroller-on-a-w1209-thermostat

**Comment by VK6TT** [comment]
Hi,

I didn't see anything to suggest the device is code compatible with the 
STM8 devices. Perhaps only having one good eye at present meant I missed 
it. However, this would make it hard for anyone on this mailing list to 
offer suggestions on how to program since we are focussed on the STM8 
series devices.

Best regards

Richard

**Comment by TG9541** [comment]
Interesting, that's the 3rd µC in a STM8S103 TSSOP20 form factor I get to know about (the others being N76E003AT20 - MCS-51 and CH32V003F4P6 - RISC-V). 

This one appears to have a very basic proprietary 8 bit core with (fixed?) 16bit instruction width:

![image](https://user-images.githubusercontent.com/5466977/229692573-76b762e2-969f-425b-b8c7-514c3da6c08c.png)

At least part of the instruction set is memory access controlled (e.g, multiplication, division):
![image](https://user-images.githubusercontent.com/5466977/229693007-2bdcb300-9832-4e16-b429-6bf8e210d72d.png)

I doubt that this chip is more than a stop-gap to serve the domestic market. I would rather not try mastering this one.

A complete datasheet is here:
https://datasheet.lcsc.com/szlcsc/1912111437_Qingdao-Eastsoft-Communication-Tech-ES7P003FGTF_C394846.pdf

**Comment by rtek1000** [comment]
I didn't expect to be able to use the same firmware, but the logic can be reused.

I am interested in using the W1209 board with DS18B20 sensors, I currently use Attiny85 with [this configuration](https://github.com/rtek1000/ATtiny85_DS18B20_Thermostat), but I have to assemble the board manually.

Incredibly ES7P003 has support in VScode ([Extension Essemi](https://marketplace.visualstudio.com/items?itemName=luh.esext)), I'm trying if I can make it work, the documents are in Chinese. I am using translator to read and access the website. :sweat_smile:

But I'm a little wary of VScode, because Platformio generated a lot more code in VScode with SDCC than STVD with Cosmic, commented [here](https://github.com/platformio/platform-ststm8/issues/60).

**Comment by TG9541** [comment]
I guess it would be worth thinking about producing an open-source STM8 board that can be ordered, e.g., from JLCPCB at a budget:

* a form factor that fits in a standard enclosure
* through-hole assembly options, e.g., for headers, relays and a LED display
* e.g., STM8SxxxKyT6C µC (e.g., STM8S103K3, STM8S903K3, STMS105K6...), OR maybe the TSSOP20 form factor so that optionally any of the other chips can be used (the CH32V003F4P6 RISC-V sure looks interesting).

**Comment by rtek1000** [comment]
Yes, I've already ordered from JLCPCB, it's a very good option.

**Comment by rtek1000** [comment]
What about the noise immunity of the STM8? Atmel's original ATmega chips were much better than Microchip's PICs.

**Comment by VK6TT** [comment]
Don't spend any time on this Thomas because I doubt the time and effort 
is warranted. It was mid 2019 when I had a derivative of the W1209 board 
made and I have never found a need to pick one up again.

A library of gerbers might be the simplest way. If anyone wants the 
gerbers, pick and place file and BOM file for this board they only need 
ask.

The board was used on top of another board containing a number of large 
power fets and was used as a voltage sensitive relay.

I only mention this because it is a real leap to go from something that 
is essentially one-off for your own use to a PCB with wider appeal. When 
I think about all the STM8 projects I have built and the absence of 
requests for a PCB or details on how to have one built I'm convinced 
your time could be better spent elsewhere. OR have you had success with 
other people wanting to emulate on of your projects?

**Comment by TG9541** [comment]
@VK6TT I don't intend to make a W1209 replacement, to be honest, the hardware feels too limited to be really fun. What I have in mind is a thing that costs about $5 in a batch of 3 pcs, good for hobby applications with a few options, e.g., RS485, data Flash, maybe RTC, one or two relays (solid state or conventional), sensors...

@rtek1000 the noise immunity of STM8 devices is good. Must be good because automotive specs aren't easy to meet.

**Comment by rtek1000** [comment]
Really limited, whoever designed the first W1209 boards could have left the UART pins for the buttons, or multiplexing the buttons with the common display pins, would help a lot to implement remote control.  I did some tests with a modified scheme, and the UART operating with an interrupt, it turned out really well.

**Comment by rtek1000** [comment]
On the issue of support for ES7P003, VScode for Linux does not find the core, they even forgot that in Linux it matters whether the file name is in uppercase.  And the Windows VScode is also missing something, it couldn't generate binary even for example code, but it's already managed to find the core.

I will try to contact the Essemi (Eastsoft) manufacturer, to correct the VScode, maybe they are always testing the changes on a computer that has already worked, so they didn't test it on a clean system.

It would also be interesting for them to make the programmer's source code available via UART, to add English language.

**Reply by TG9541** [reply]
Myself I would consider the ES8P003 based board a loss. I don't think that in the presence of the CN32V003, a quite powerful device, that PIC knock-off has much of a future, even on the Chinese market.

**Comment by rtek1000** [comment]
About Attiny85, I already started designing the custom board, in Kicad, I'm going to use SMD. When I mount manually, I use almost everything PTH.

**Comment by rtek1000** [comment]
On the issue of support for ES7P003:

- VCcode was able to generate the Hex file, other suggested extensions were interfering with the Essemi extension
- - I left only the 3 extensions installed:
- - - Chinese (Simplified) (????) Language Pack for Visual Studio Code
- - - ES Extensions
- - - essemi-language-server
- - Need to install other things like core separately.

Still missing the part of programming the uC, I'll try to use a serial port and/or J-link V8.

I found these tips (Strangely, the blog's website did not allow google translator to operate):

> There are two development environments iDesigner4.1 and VSCode For ESSEMI, in
> It can be downloaded at http://www.essemi.com/article/c80/p2.html?searchWord=32%E4%BD%8D. You need to install iDesigner and vscode first in "Pre-installation Environment" and "C Compiler Toolchain HRCC (8-bit)".
> ESBurner is used to configure the configuration words of the chip, including clock, simulation, reset, IO and other options.
> Emulator: es-link II, can emulate and download 8-bit and 32-bit chips.
> There are a few places that need special attention:
> When simulating, you need to open ESBurner, select the device (update chip timing)-chip-configuration, select ICDEN to enable, external reset needs a reset circuit, and select IO as internal reset. Simulation is required to close ESBurner. It is very important, otherwise it cannot be simulated.
> When ICDEN is enabled, the program cannot be burned into the chip. ICDEN needs to be banned. Very important.
> ———————————————
> Copyright statement: This article is an original article by CSDN blogger "Clodiman", and follows the CC 4.0 BY-SA copyright agreement. Please attach the original source link and this statement for reprinting.
> Original link: https://blog.csdn.net/weixin_48142918/article/details/118089007

**Comment by rtek1000** [comment]
On the issue of support for ES7P003:

- I was not able to program the ES7P003FGTF with J-Link, the manufacturer's application does not support this programmer. And the J-Link programmer app does not support this manufacturer.

- I couldn't find the ES-Link programmer to buy on Aliexpress, where W1209 boards are usually sold in bulk.

![No Jlink support](https://user-images.githubusercontent.com/8660811/230029848-484ebed7-737a-4f08-8873-9e43ca753ca1.png)

- The UART-BOOT application apparently needs a bootloader that the user must install on the microcontroller in a similar way to what is done with Microchip's PIC microcontrollers.

- I didn't find anything about Boot mode (bootloader) in the datasheet.

- In the datasheet it still says that the UART is limited to 38400 bauds (UART The transmission baud rate is less than 38400 bps, page 136 of the datasheet in 6.4.7 UARTsPrecautions for use).

- I couldn't access the manufacturer's forum, because you need a login, and login validation is done with SMS and only accepts a mobile number from China, but the registration form does not accept a country code.

- Contact with the company is only via telephone or physical address, there is no email or contact form on the website.

- In short, it doesn't seem to have much desire to cater to people on the west side at the moment.

- In case anyone wants to avoid this board, note the absence of the Vcap capacitor near the microcontroller.

- The board also lacks the NTC filter, which might need further software handling.

![W1209_no_STM8S003](https://user-images.githubusercontent.com/8660811/230031762-3cd1df96-8cd0-4875-bcc5-88c40d079341.jpg)

**Reply by TG9541** [reply]
All the STM8S103F3P6-look-alikes I'm aware of (N76E003AT20, CH32V003F4P6 and ES7P003FGTF) use the VCAP pin for some other function. Not populating the 1µF VCAP is a frequent practice. I've seen one board ([XY-LPWM](https://github.com/TG9541/XY-LPWM)) that clearly had been redesigned for *N76E003AT20 only*: no VCAP pads but an extra header pin for ICP instead.

Not populating the NTC cap can be a feature: if the noise is just right this can help avoid rectifying noise current (-> offset). That way software filters can turn noise into digits (check out the usage of `lpf32` in [this code](https://github.com/TG9541/W1209/blob/master/measure.fs)).  
 
Your experience with the ES7P003FGTF tool chain is telling: this is bottom-drawer grade engineering. Likely a very small team that designed the chip and the ISA around readily available design blocks and bits and pieces of tool chains. The specs are likely such that the chip can be used for some applications if performance and long term availability is very low on the feature list.
